### ISSUE TITLE
Implement COALESCE function

### ISSUE THREAD

**Issue by malin1993ml** [cmu-db/peloton#563]
Here's what we get now:
```sql
postgres=# create table a(id int primary key, value varchar);
create table a(id int primary key, value varchar) 0
postgres=# insert into a values(1, 'hi');
insert into a values(1, 'hi') 1
postgres=# select * from a;
 id | value 
----+-------
  1 | hi
(1 row)

postgres=# select coalesce(id, value) as coal from a;
function coalesce not found.
```

It seems that we already have the function type:
https://github.com/cmu-db/peloton/blob/master/src/include/type/types.h#L227
But there's no implementation for that.

**Comment by sumanthrao** [comment]
I would like to work on this issue,

**Comment by apavlo** [comment]
@sumanthrao Thanks for the offer. This blocked on #838 though. Once we get the new built-in function infrastructure implemented, then this will definitely be doable.

**Comment by poojanilangekar** [comment]
@sumanthrao Are you still interested in working on this? PR #838 has been merged and so it should be pretty straightforward to implement the COALESCE function.

**Comment by schedutron** [comment]
@poojanilangekar May I work on this?

**Comment by poojanilangekar** [comment]
@schedutron I was just cleaning up old issues in December. It is best you ask @pervazea if anyone is working on this.

**Comment by schedutron** [comment]
@pervazea May I work on this issue?

**Comment by pervazea** [comment]
Go ahead.

**Comment by schedutron** [comment]
What should be the return type of `coalesce`? `UDT`?

**Comment by pervazea** [comment]
According to the Postgres documentation, coalesce returns the first non-null argument. The type will be whatever the type is of that argument. You should be able to exercise the function in Postgres, and and see what it does.

One of the experts should correct me if I'm reading it wrong.

**Comment by schedutron** [comment]
Yes, that could be any type, so with what return type should I register it in `AddBuiltinFunction` in `catalog.cpp`?

**Comment by schedutron** [comment]
`COALESCE()` is a function which takes a variable number of arguments. A similar function is `CONCAT()`; I looked at its implementation for inspiration but that didn't help since the current implementation concatenates just two columns. So how do I register `COALESCE()` as a function that can take a variable number of arguments, in `catalog.cpp`? Specifically, what should be the second argument passed to [`AddBuiltinFunction()` in `catalog.cpp`](https://github.com/cmu-db/peloton/blob/c15635bd9da44214a9bb4498e4b7bd4b7928d552/src/catalog/catalog.cpp#L829), which is a vector representing the types of the arguments which the function can take?

As a starting step, should I too should implement `COALESCE()` for just 2 columns?

**Comment by schedutron** [comment]
@poojanilangekar `COALESCE()` is a function which takes a variable number of arguments. A similar function is `CONCAT()`; I looked at its implementation for inspiration but that didn't help since the current implementation concatenates just two columns. So how do I register `COALESCE()` as a function that can take a variable number of arguments, in `catalog.cpp`? Specifically, what should be the second argument passed to [`AddBuiltinFunction()` in `catalog.cpp`](https://github.com/cmu-db/peloton/blob/c15635bd9da44214a9bb4498e4b7bd4b7928d552/src/catalog/catalog.cpp#L829), which is a vector representing the types of the arguments which the function can take?

As a starting step, should I too should implement `COALESCE()` for just 2 columns?
